### PR TITLE
[RFC] Splitting up react-vis into separate packages

### PR DESCRIPTION
**Pull request by schnerd**
> **Note:** This PR should serve as a proof-of-concept and forum for discussion, and is not intended to be merged in it's current state.

### Motivation

Over time react-vis has grown to support all sorts of functionalities–canvas-based charts, webgl charts at one point, and many advanced chart types like sankey, sunbursts, contours, custom svg series, etc. While this flexibility is great, many of these features are less well supported than the core chart types and are often the focus of many issues in github, leading to more maintenance overhead for contributors.

Moving forward, we propose splitting up react-vis into separate packages (in separate repos). The core `react-vis` package will still contain all the common chart types and utilities people use, but we will start publishing more advanced features as separate packages such as `react-vis-sankey`, `react-vis-sunburst`, `react-vis-custom-svg-series`, etc. (could even be referred to as 'plugins' if we wanted). We hope that this will have the following impact:

- Better expectations on what features are robust vs experimental. The core `react-vis` experience should be smooth, consistent, bug-free, but external packages may have some rough edges (and can be labeled as such in README.md).
- Ability to distribute ownership/maintenance to the community. It's hard for us to maintain dozens of chart types, but seeking external maintainers for some packages may empower our community and create a vibrant ecosystem of packages/plugins.

While it's possible to publish multiple packages from a single monorepo, creating multiple repos seems to be the strategy that best achieves the goals above.

### Potential Issues

There are potential downsides to splitting things up into packages:

1. Issues may be spread across many repos now. That said, for repos still primarily maintained by Uber, we can optionally direct users to add issues to the main `react-vis` repo instead.

2. More internal helpers exported from `react-vis` – theme constants, layout utils (like getInnerDimensions), etc may need to be exported so the other chart/series types can access them. These should probably be documented as well if they're exported.

3. Overhead of creating new repo and boilerplate to build new experimental features.

While these are potential issues, we hope that the pros outweigh the cons.

### Technical Implementation
In the first pass, we propose creating the following packages by extracting the associated code out of the core react-vis repo:

```
react-vis-whisker-series
react-vis-radar-chart
react-vis-sankey
react-vis-sunburst
react-vis-treemap
react-vis-parallel-coordinates
react-vis-contour-series
react-vis-custom-svg-series
react-vis-heatmap-series
react-vis-polygon-series
```

Later we may consider extracting canvas charts into `react-vis-canvas`, and also extracting other packages such as `react-vis-decorative-axis`, etc.

The code in the current PR is a proof-of-concept for splitting the `sankey` chart out into a separate package. The new package can be found as a PR on the following temporary repo: https://github.com/schnerd/playground/pull/1

- New packages have a peerDependencies on `react`, `prop-types`, and `react-vis`, and also includes them as devDependencies for running unit tests, etc.
- The new package can import various utils from the core react-vis package (Theme.DISCRETE_COLORS, getInnerDimensions, etc)
- If we want to continue to document & showcase the new packages on the react-vis website, we will add `react-vis-*` as a dep in either `/showcase` or `/website`.
- New packages will be published under the uber (or uber-web) github org, however for certain repositories we will advertise that we are looking for external maintainers, and we should be open to transferring repository ownership to community members that have demonstrated strong ownership (need to figure out licensing issues here)

### Next Steps

Before we move forward, it would be good to get your thoughts on this proposal. Are there issues with the proposed technical implementation? Do the pros outweigh the cons for this new setup? If there's general agreement, I'll likely carry out the implementation work.

### CODE REVIEW

File: src/animation.js [hunk 1]
```diff
@@ -54,6 +69,8 @@ export function extractAnimatedPropValues(props) {
 }
 
 class Animation extends PureComponent {
+  static ANIMATED_SERIES_PROPS = ANIMATED_SERIES_PROPS;
```
Comment (schnerd) [inline src/animation.js]: Seemed like having this as a static property on `Animation` instead of a util exported from `series-utils` was a cleaner API for access from other packages.

Comment (ibgreen) [reply src/animation.js]: This looks exciting!

FWIW, in deck.gl etc we are also introducing multiple npm modules to allow users to select the pieces that they need, but we are stil keeping all the code inside a single github repository (them in separate source trees). We prefer this approach for a couple of reasons. In particular, we feel that there is a significant overhead in keeping build tooling, scripts, techniques etc consistent across multiple repos.

**Comment by mcnuttandrew** [comment]
Interesting work! Pulling the library apart into separate npm modules seems like a great idea, but I have some worries about doing so in the manner described.

- Getting folks to support things/components in the library is already pretty difficult, it seems like if the library was fragmented into many repos that would become more difficult.
- A constellation of repos would make it hard to make full library changes (I remember there was some talk of a sweeping API change/audit a while ago?)
- I am curious if there is any available data on which charts/series types are used and to what degree, or even more which sets of things tend to be used together. This data would enable an informed choice about what should and should not be in the core library.
- I'm not sure there is a substantial enough contributing community warrant transferal ownership of sub-packages. It 100% could be the case that advertising for maintainers would lead to maintainers, but I worry it would just lead to those features becoming abandonware, damaging the value proposition of react-vis in total.

The TLDR of my comments is that I agree with the spirit of Ib's comment: i think a mono repo would be a better approach.

**Comment by ericsoco** [comment]
The value of publishing multiple npm packages from a single monorepo is primarily about bundle size for end users, and possibly also separation of concerns and legibility across the whole codebase.

The value of a monorepo, regardless of how concerns are modularized, is everything @ibgreen and @mcnuttandrew mentioned -- tooling, consistency, centralized identity. One problem I've seen consistently with projects split across multiple sub-repos (e.g. [Vega](https://github.com/vega/vega)) is that it's confusing where to find answers (in the source) and where to ask questions (file issues). I've also found that it doesn't help distribute maintenance; on the contrary, the proliferation of sub-repos can cause confusion among potential contributors.

I think deck.gl has a pretty good strategy for separating core from layers from experimental layers; all are contained within a single repo (@ibgreen correct me if I'm wrong, been a bit since I've contributed) but boxed and labeled clearly. Utils exist as separate packages from separate repos when their utility could be of value beyond deck.gl (e.g. probe.gl, mjolnir, etc).

If we decide that modularity and bundle size is a motivating concern for a reorg, my vote would be for publishing modules from a monorepo AND putting a lot of effort into simplifying and clarifying the contribution process. The latter could help to address the goal of distributing maintenance without fragmenting the project and creating abandonware.

My two cents...but I'm not a heavy contributor to react-vis so take it at face value.

**Comment by schnerd** [comment]
>The value of publishing multiple npm packages from a single monorepo is primarily about bundle size for end users, and possibly also separation of concerns and legibility across the whole codebase.

Bundle size should be less an issue moving forward as bundlers continue to improve their tree-shaking support. And I think "separation of concerns and legibility" can be achieved through good directory hierarchy and documentation, without the need to actually publish multiple modules.

**Comment by schnerd** [comment]
Appreciate the input all, sounds like there's general agreement that numerous repos would have quite a few drawbacks and should be avoided. It seems like the best path forward is either to deprecate features we no longer want to support, or have a maximum of two repos (core + extended). I'll sync with @jckr to get his thoughts.
